django/db/models/sql/query.py

```diff
@@ -1581,13 +1604,17 @@ def add_filtered_relation(self, filtered_relation, alias):
                 "FilteredRelation's relation_name cannot contain lookups "
                 "(got %r)." % filtered_relation.relation_name
             )
-        for lookup in chain(lookups):
-            lookup_parts, lookup_field_parts, _ = self.solve_lookup_type(lookup)
-            shift = 2 if not lookup_parts else 1
-            lookup_field_path = lookup_field_parts[:-shift]
+        for lookup in get_children_from_q(filtered_relation.condition):
+            _, lookup_field_parts, _ = self.solve_lookup_type(lookup)
+            lookup_field_path = lookup_field_parts[:-1]
+            intersects = (
+                lookup_field_path
+                and relation_field_parts
+                and lookup_field_path[0] == relation_field_parts[0]
+            )
```

Comment:
Not entirely sold on the necessity of this change in this PR.

What I meant by https://github.com/django/django/pull/16622#issuecomment-1529061058 is that `test_conditional_expression_with_func` should be adapted not to require adjusting this logic.

In other words, I believe there are ways to test that `FilteredRelation` doesn't _crash_ on conditional expressions (which is the purpose of this PR) without getting into adjusting what constitutes a violation? Maybe I'm missing something though.

Reply:
@charettes Thanks for the explanation. We could solve this is avoiding to check paths in the rhs of expressions, and changing `test_conditional_expression_with_func` to not assumine that queryset evaluation raise a `ValueError`.

Checking rhs  was the matter of [this comment](https://github.com/django/django/pull/16622#issuecomment-1455677532) by @zhu , but probably we could postpone this in a future PR, as you suggested.